### PR TITLE
issue-43: support regex in writeable database whitelist configuration

### PR DESCRIPTION
**Pull request by courtsVII**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ## TBD
 ### Added
 * Configurable SSH session timeout for SSH tunnels.
+* Wild Cards enabled in writeable database whitelist
```
Comment (pradeepbhadani) [inline CHANGELOG.md]: writeable --> writable

File: waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/server/security/DatabaseWhitelistAccessControlHandlerTest.java [hunk 2]
```diff
@@ -46,7 +46,7 @@
   @Mock
   private FederationService federationService;
   private DatabaseWhitelistAccessControlHandler handler;
-  private final List<String> whitelist = newArrayList("writabledb");
+  private final List<String> whitelist = newArrayList("writabledb", "regexdb*");
```
Comment (pradeepbhadani) [inline waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/server/security/DatabaseWhitelistAccessControlHandlerTest.java]: Can you add `*` in whitelist for testing?

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/security/DatabaseWhitelistAccessControlHandler.java [hunk 3]
```diff
@@ -41,15 +45,35 @@ public DatabaseWhitelistAccessControlHandler(PrimaryMetaStore primaryMetaStore,
...
+    if (databaseWhiteList.contains(databaseName)) {
+      return true;
+    }
+
+    char firstLetter = databaseName.charAt(0);
+    char secondLetter = (char) ((char) firstLetter + 1);
+    Set<String> sortedSubSet = databaseWhiteList.subSet(String.valueOf(firstLetter), String.valueOf(secondLetter));
+    for (String whiteListEntry : sortedSubSet) {
+      if (whiteListEntry.endsWith("*")) {
+        Matcher matcher = Pattern.compile(cleanString(whiteListEntry)).matcher(databaseName);
```
Comment (pradeepbhadani) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/security/DatabaseWhitelistAccessControlHandler.java]: This works if input is only `*` ?

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,6 +1,7 @@
 ## TBD
 ### Added
 * Configurable SSH session timeout for SSH tunnels.
+* Wild Cards enabled in writable database whitelist
```
Comment (massdosage) [inline CHANGELOG.md]: .

File: README.md [hunk 5]
```diff
@@ -124,6 +124,7 @@ Example:
       writable-database-white-list:
       - my_writable_db1
       - my_writable_db2
+      - user_db_*
```
Comment (massdosage) [inline README.md]: We should document this better, probably add a section in the whitelisting describing that you can either do full db names or use reg ex and then link to the Java Pattern documentation https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/security/DatabaseWhitelistAccessControlHandler.java [hunk 6]
```diff
@@ -41,15 +45,36 @@ public DatabaseWhitelistAccessControlHandler(PrimaryMetaStore primaryMetaStore,
...
     if (databaseName == null) {
       return true;
     }
-    return databaseWhiteList.contains(databaseName.trim().toLowerCase());
+    databaseName = trimToLowerCase(databaseName);
+    if (databaseWhiteList.contains(databaseName) || databaseWhiteList.contains("*")) {
+      return true;
+    }
+
+    char firstLetterOfDatabaseName = databaseName.charAt(0);
```
Comment (massdosage) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/security/DatabaseWhitelistAccessControlHandler.java]: Aah, I guess this is back to "WIP" now seeing as we decided on full regex right? So I'll stop reviewing for now...

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,6 +1,7 @@
 ## TBD
 ### Added
 * Configurable SSH session timeout for SSH tunnels.
+* Regexes enabled in writable database whitelist
```
Comment (massdosage) [inline CHANGELOG.md]: FULL STOP

File: README.md [hunk 8]
```diff
@@ -146,7 +147,7 @@ The table below describes all the available configuration values for Waggle Danc
...
 | `primary-meta-store.access-control-type`              | No       | Sets how the client access controls should be handled. Default is `READ_ONLY` Other options `READ_AND_WRITE_AND_CREATE`, `READ_AND_WRITE_ON_DATABASE_WHITELIST` and `READ_AND_WRITE_AND_CREATE_ON_DATABASE_WHITELIST` see Access Control section below. |
-| `primary-meta-store.writable-database-white-list`     | No       | White-list of databases used to verify write access used in conjunction with `primary-meta-store.access-control-type`. The list of databases should be listed without any `primary-meta-store.database-prefix`. |
+| `primary-meta-store.writable-database-white-list`     | No       | White-list of databases used to verify write access used in conjunction with `primary-meta-store.access-control-type`. The list of databases should be listed without any `primary-meta-store.database-prefix`. This property supports both full database names and RegEx patterns. Further information on Java RegEx patterns can be found [here](https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html)|
```
Comment (massdosage) [inline README.md]: Full stop

Comment (massdosage) [reply README.md]: and [Java RegEx patterns](https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html).|

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/security/DatabaseWhitelistAccessControlHandler.java [hunk 9]
```diff
@@ -41,15 +45,30 @@ public DatabaseWhitelistAccessControlHandler(PrimaryMetaStore primaryMetaStore,
...
   }
 
   @Override
   public boolean hasWritePermission(String databaseName) {
     if (databaseName == null) {
       return true;
     }
-    return databaseWhiteList.contains(databaseName.trim().toLowerCase());
+    databaseName = trimToLowerCase(databaseName);
+    if (databaseWhiteList.contains(databaseName)) {
```
Comment (massdosage) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/security/DatabaseWhitelistAccessControlHandler.java]: Check whether this if statement is ever returning true for tests, I suspect not and if so then it can be removed

File: waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/server/security/DatabaseWhitelistAccessControlHandlerTest.java [hunk 10]
```diff
@@ -46,7 +46,7 @@
   @Mock
   private FederationService federationService;
   private DatabaseWhitelistAccessControlHandler handler;
-  private final List<String> whitelist = newArrayList("writabledb");
+  private final List<String> whitelist = newArrayList("writabledb", "regexdb.*");
```
Comment (massdosage) [inline waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/server/security/DatabaseWhitelistAccessControlHandlerTest.java]: regexdb -> user.* (to be more like requirements)

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/security/DatabaseWhitelistAccessControlHandler.java [hunk 11]
```diff
@@ -16,21 +16,23 @@
...
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 import com.hotels.bdp.waggledance.api.federation.service.FederationService;
 import com.hotels.bdp.waggledance.api.model.PrimaryMetaStore;
 
 public class DatabaseWhitelistAccessControlHandler implements AccessControlHandler {
 
-  private final Set<String> databaseWhiteList = new HashSet<>();
+  private final List<Pattern> databaseWhiteList = new ArrayList<>();
```
Comment (massdosage) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/security/DatabaseWhitelistAccessControlHandler.java]: Why the change from Set to List? Duplicates don't really make sense here so I think a Set is better.
